### PR TITLE
run,mount: remove path only if they didnt pre-exist

### PR DESCRIPTION
**Pull request by flouthoc**
It seems buildah was cleaning path after unmounting the content which was added by `--mount`, however buildah should `remove` only if the path it did not pre-exist otherwise it should just simply `unmount`.

Following behavior aligns `buildah `to `buildkit`

Closes: https://github.com/containers/buildah/issues/4739

#### What type of PR is this?

> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test 
> /kind feature
> /kind flake
> /kind other

#### Does this PR introduce a user-facing change?

```release-note
run,mount: remove path only if they didnt pre-exist
```

### CODE REVIEW

File: run_common.go [hunk 1]
```diff
@@ -1464,10 +1464,28 @@ func cleanableDestinationListFromMounts(mounts []spec.Mount) []string {
 	return mountDest
 }
 
+func checkIfMountDestinationPreExsits(root string, dest string) (bool, error) {
+	statResults, err := copier.Stat(root, "", copier.StatOptions{}, []string{dest})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (flouthoc) [reply run_common.go:1469]: @rhatdan Did you mean `ENOENT`, yes it does in stat results as an `Error` string.

**Comment by flouthoc** [comment]
Not sure why conformance is timing out,let me check if its happening in other PR's as well or not.

**Comment by flouthoc** [comment]
Created PR for conformance test: https://github.com/containers/buildah/pull/4760

File: run_common.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhatdan) [reply run_common.go:1469]: Yes

[the remark this answers is not in the document]
Comment (rhatdan) [reply run_common.go:1469]: My point being would this fail if the path does not exist? 

Should we have 
if err != nil {
Not return an error if ENOENT?

[the remark this answers is not in the document]
Comment (flouthoc) [reply run_common.go:1469]: @rhatdan No this does not returns `ENOENT` here, that happens as part of statResults here error is related to the copier itself not about `stat`inside the root directory, I verified it.

[the remark this answers is not in the document]
Comment (nalind) [reply run_common.go:1469]: Typo in function name.

File: tests/bud.bats [hunk 2]
```diff
@@ -5693,6 +5693,18 @@ _EOF
...
+@test "bud-verify-if-we-dont-clean-prexisting-path" {
+  skip_if_no_runtime
+  skip_if_in_container
+  run_buildah 1 build -t testbud $WITH_POLICY_JSON -f $BUDFILES/verify-cleanup/Dockerfile $BUDFILES/verify-cleanup/
+  expect_output --substring "hello"
+  expect_output --substring "Directory /tmp exists."
+  expect_output --substring "Directory /var/tmp exists."
+  expect_output --substring "Directory /testdir DOES NOT exists."
+  expect_output --substring "/tmp/hey: No such file or directory"
+
```
Comment (nalind) [inline tests/bud.bats:5705]: Extra blank line.

**Comment by nalind** [comment]
What's the rationale for only checking these things for specific types of mounts?

**Comment by flouthoc** [comment]
> What's the rationale for only checking these things for specific types of mounts?

Should happen for all the mounts now, I added test to verify this in few of them.

**Comment by TomSweeneyRedHat** [comment]
Needs a rebase @rhatdan

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
@nalind @TomSweeneyRedHat @vrothberg @giuseppe @umohnani8 PTAL

File: run_common.go [hunk 3]
```diff
@@ -1497,63 +1520,71 @@ func (b *Builder) runSetupRunMounts(mounts []string, sources runMountInfo, idMap
...
+			finalMounts = append(finalMounts, *mountSpec)
 			if tl != nil {
 				targetLocks = append(targetLocks, tl)
 			}
 		default:
 			return nil, nil, fmt.Errorf("invalid mount type %q", mountType)
 		}
+
+		if mountSpec != nil {
+			pathPrexists, err := checkIfMountDestinationPreExists(mountPoint, mountSpec.Destination)
```
Comment (umohnani8) [inline run_common.go:1577]: nit: variable name typo s/pathPrexists/pathPreExists

**Comment by TomSweeneyRedHat** [comment]
Once the nits are squished, LGTM

File: run_common.go (hunk 3 above)
Comment (flouthoc) [reply run_common.go]: Fixed.

**Pull request merged**